src/contracts/tokens/RewardPool.sol

```diff
@@ -0,0 +1,237 @@
+            _getReward(msg.sender);
+        }
+    }
+
+    /// @inheritdoc IRewardPool
+    function getReward() external updateReward(msg.sender) isAuthorized {
+        _getReward(msg.sender);
+    }
+
+    function _getReward(address _account) internal {
+        uint256 _reward = earned(_account);
+        if (_reward > 0) {
+            rewards = 0;
+            rewardToken.safeTransfer(_account, _reward);
+            emit RewardPool_RewardPaid(_account, _reward);
+        }
+    }
+
+    /// @inheritdoc IRewardPool
+    function rewardPerToken() public view returns (uint256) {
+        if (totalStaked() == 0) return rewardPerTokenStored;
+        uint256 _timeElapsed = lastTimeRewardApplicable() - lastUpdateTime;
+        return
+            rewardPerTokenStored +
+            ((_timeElapsed * rewardRate * 1e18) / totalStaked());
+    }
+
+    /// @inheritdoc IRewardPool
+    function earned() public view returns (uint256) {
+        return
+            ((_totalStaked * (rewardPerToken() - rewardPerTokenPaid)) / 1e18) +
+            rewards;
+    }
+
+    /// @inheritdoc IRewardPool
+    function queueNewRewards(uint256 _rewardsToQueue) external isAuthorized {
+        uint256 _totalRewards = _rewardsToQueue + queuedRewards;
+
+        if (block.timestamp >= periodFinish) {
+            notifyRewardAmount(_totalRewards);
+            queuedRewards = 0;
+            return;
+        }
+
+        uint256 _elapsedTime = block.timestamp -
+            (periodFinish - _params.duration);
+        uint256 _currentAtNow = rewardRate * _elapsedTime;
+        uint256 _queuedRatio = (_currentAtNow * 1000) / _totalRewards;
+
+        if (_queuedRatio < _params.newRewardRatio) {
+            notifyRewardAmount(_totalRewards);
+            queuedRewards = 0;
+        } else {
+            queuedRewards = _totalRewards;
+        }
+    }
+
+    /// @inheritdoc IRewardPool
+    function notifyRewardAmount(
+        uint256 _reward
+    ) public updateReward(address(0)) isAuthorized {
+        if (_reward == 0) revert RewardPool_InvalidRewardAmount();
+        if (block.timestamp >= periodFinish) {
+            rewardRate = _reward / _params.duration;
+        } else {
+            uint256 _remaining = periodFinish - block.timestamp;
+            uint256 _leftover = _remaining * rewardRate;
+            rewardRate = (_reward + _leftover) / _params.duration;
+        }
+
+        lastUpdateTime = block.timestamp;
+        periodFinish = block.timestamp + _params.duration;
+        emit RewardPool_RewardAdded(_reward);
+    }
+
+    /// @inheritdoc IRewardPool
+    function emergencyWithdraw(
+        address _rescueReceiver,
+        uint256 _wad
+    ) external isAuthorized {
+        if (_wad == 0) revert RewardPool_WithdrawNullAmount();
+        IERC20(rewardToken).safeTransfer(_rescueReceiver, _wad);
+        emit RewardPool_EmergencyWithdrawal(msg.sender, _wad);
+    }
+
+    // --- Modifiers ---
+
+    modifier updateReward(address _account) {
```

Comment:
if the `_account` is meant to be RewardManager every time, I wouldnt use arbitrary accounts here. Just use the rewards manager directly.